crates/fuel-core/src/graphql_api/ports.rs

```diff
@@ -130,3 +149,38 @@ pub trait DatabaseChain {
 
     fn base_chain_height(&self) -> StorageResult<DaBlockHeight>;
 }
+
+pub trait FindTx {
+    fn find_one(&self, id: TxId) -> Option<TxInfo>;
+}
+
+pub trait InsertTx {
+    fn insert(&self, txs: Vec<Arc<Transaction>>) -> Vec<anyhow::Result<InsertionResult>>;
+}
+
+pub trait TxSubscription {
+    fn tx_update_subscribe(&self) -> tokio::sync::broadcast::Receiver<TxUpdate>;
+}
+
+pub trait TxPoolPort: Send + Sync + FindTx + InsertTx + TxSubscription {}
+
+#[async_trait]
+pub trait DryRunExecution {
+    async fn dry_run_tx(
+        &self,
+        transaction: Transaction,
+        height: Option<BlockHeight>,
+        utxo_validation: Option<bool>,
+    ) -> Result<Vec<Receipt>>;
+}
+
+pub trait BlockProducerPort: Send + Sync + DryRunExecution {}
+
+pub trait ExecuteWithoutCommit {
```

Comment:
`Database` should be generic of the trait

Reply:
I added it as a generic, but I'm not sure what else you wanted this to entail, so you could please let me know if I didn't everything related you wanted?